### ISSUE TITLE
Add a maxValue and minValue to int

### ISSUE THREAD

**Issue by gspencergoog** [dart-lang/sdk#41717]
I would like to request adding constant values to `int` called (for instance) `maxValue` and `minValue` that represent the largest and smallest valid integers.

It would be helpful to be able to ask the Dart runtime the largest (and smallest) number an `int` can represent, since it depends on the platform that it is running on.

For instance, if I wanted to write an integer constraint class similar to Flutter's `BoxConstraints`, I wouldn't know what to pass for an unconstrained value (the way it passes `double.infinity`). Same thing if I want to write generic logic that uses an integer `min` or `max` function where I sometimes want it unbounded (or as unbounded as a limited precision int can be).

Sure, I can use `num`, but then I'm not really using integers all the time, and it would be nice to know that I can stick to precise values.

**Comment by lrhn** [comment]
The values would be `-0x8000000000000000` and `0x7fffffffffffffff` for native code and `-double.infinity` and `double.infinity` for JS compiled code. Those are the minimal and maximal values allowed by the `int` type on those platforms.

I don't recommend representing "no limit" with an actual limit. I'd rather use `null` for that, and then have helper functions like:
```dart
int? maybeMin(int? value1, int? value2) => 
   (value1 != null && (value2 == null || value2 > value1)) ? value1 : value2;
```

If you want the limits, they are easily defined:
```dart
final int maxInt = (double.infinity is int) ? double.infinity as int : ~minInt;
final int minInt = (double.infinity is int) ? -double.infinity as int : (-1 << 63);
```

**Comment by gspencergoog** [comment]
Well, but using the actual limits in the double case is so nice and understandable.  Here's a clamp function with optional limits, for instance:

```dart
double clamp(
  double value, {
  double min = -double.infinity,
  double max = double.infinity,
}) => value.clamp(min, max);
```

It's trivial to understand that the defaults indicate that the call is unbounded by default.

As for using `null` for "no limit", what if I also need to represent a negative "no limit" as above? And `null` is also overloaded with "didn't provide a value" in many cases.

Thanks for those definitions, however, they could be helpful. They're pretty non-obvious though: It wouldn't occur to me that a `double.infinity` would ever be an `int`, seeing as how it starts off as a double.

Could those expressions be `const`, or would they not work correctly in all cases?

**Comment by jamesderlin** [comment]
> The values would be `-0x8000000000000000` and `0x7fffffffffffffff` for native code and `-double.infinity` and `double.infinity` for JS compiled code. Those are the minimal and maximal values allowed by the `int` type on those platforms.

While technically correct that the "maximum" and "minimum" values would be `double.infinity` and `-double.infinity` for JavaScript code, what most people *want* are constants for (2^53 - 1) and (-2^53 + 1): the limits at which `int`s stop behaving like typical `int`s and might no longer be exact.

**Comment by lrhn** [comment]
What limit you want depends on what you are going to use it for.
A `min` or `max` function default value could reasonably use infinity.

A limit to avoid loss of precision makes more sense when you want to iterate.

The "limit of representability" can be 2^53 or 2^53 - 1. Both are exactly representable as doubles, and so are all lower positive integers. The first unrepresentable integer is 2^53 +1.
I guess the reason for picking the (arguably too small) 2^53-1 as "limit" is that
```dart
for (var i = 0; i < limit; i++) { ... }
```
will terminate. If you use 2^53, then it won't.

**Comment by rakudrama** [comment]
TL;DR:  Use `double.maxFinite` rather than `double.infinity`.
It is the largest value you can do arithmetic on and get a useful result.

There is another definition of what it means to be an integer, given by the JavaScript `Number.isInteger` predicate. This _excludes_ infinities.

We are considering adopting this definition:
- It is better aligned with JavaScript
- `Number.isInteger` is specially implemented / understood by the JavaScript engines (V8's `Number.isInteger(x)` is faster than our current `x is int` test).

Adopting that definition would mean that `+` is not closed on `int`. I'm not sure that this is a problem, since you have to go far outside compatibility with the VM to get an infinity. I suspect you are much more likely to experience inconsistent rounding from a temporary excursion outside the 2<sup>53</sup> range. You can also get `NaN` from integer arithmetic on the web if you try hard enough and we ignore that. The infinities are mostly sticky. If we really valued mathematical consistency we would not have adopted 64-bit ints on the VM.

@lrnn It is difficult to correctly use limits. The loop with terminate for either bound, unless you pick 2<sup>53</sup> _and_ use `<=`. One generally tests an inclusive limit with `<=`, but testing against the limit is broken on the VM too, since `0x7fffffffffffffff + 1 < 0`, or said differently, `anInt <= 0x7fffffffffffffff` is always true.

There are many 'maximum' values. Should we have a constant for each use case?
- The value `x` for which there is no other value `y` such that `x < y`.
- The greatest value `x` for which `x < x + 1` (`0x7ffffffffffffffe` on the VM).
- The number of bits that work with bitwise operations.

One way to write portable code is to provide values that characterize the platform and let the developer use these values to write code that is correct under the parameterization. This can be hard work, since limits are hard to use a parameterized program is really two programs.

Another way to write portable code is write code works correctly on all platforms, i.e. a lowest common denominator approach.

Should we provide constants and predicates to support both approaches?

**Comment by gspencergoog** [comment]
Oh, I didn't know about `double.maxFinite`, that seems useful.

Looks like that is defined as `1.7976931348623157e+308`, however which means it's well outside the "accurate" (2<sup>53</sup>) integers on the web.

I resonate with @jamesderlin's [comment](https://github.com/dart-lang/sdk/issues/41717#issuecomment-622479244) above, because I think that's what I'm wanting more often out of an int limit than the "infinity" case.

It depends a lot on what you're using ints for, but for many of those use cases, if a double was good enough, I probably would use a double. Since the reason I'm using an int is that a double _isn't_ good enough, I probably don't want to stray outside of the range where it causes problems that it's represented as a double on the web. Since that range is different on different platforms, if I want to write code sensitive to those limits, it would be nice to be able to just refer to an sdk-defined constant so that I know that the SDK has my back and will give me the right values, even if I start running on a new platform or compiler/language spec version.

**Comment by bernaferrari** [comment]
I wish I had access to `max32Float` because that's what Skia uses and I can't clampDouble without that value.

**Comment by lrhn** [comment]
```dart
const maxFloat32 = 3.4028234663852886e+38;
```

There are lots of interesting values, as float-64 and float-32.
Here's a handful of the ones I usually try to use in tests, for coverage.
```dart
/// Numerically smallest non-zero 64-bit float value.
const float64MinPositive = 5e-324;
/// Largest positive denormal 64-bit float value.
const float64MaxDenormal = 2.225073858507201e-308;
/// Smallest positive normal (non-denormal) 64-bit float value.
const float64MinNormal = 2.2250738585072014e-308;
/// Largest 64-bit float value below 1.0.
const float64MaxFraction = 0.9999999999999999;
/// Largest non-integer 64-bit float value.
const float64MaxNonInteger = 4503599627370495.5;
/// Largest exact 64-bit float integer value.
/// 
/// The largest integer 64-bit float where all smaller integers 
/// are also 64-bit floats.
/// All larger doubles represent ranges larger than one integer.
const float64MaxExactInteger = 9007199254740992.0;
/// Largest finite 64-bit float value.
const float64MaxFinite = 1.7976931348623157e+308;
/// Positive infinity.
const float64Infinity = 1.0 / 0.0;
/// Not a number
const float64NaN = 0.0 / 0.0;
/// A round trap 64-bit float value at the edge of the precision of 1.0. 
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float64RoundTrap1 = 0.49999999999999994;
/// A round trap 64-bit float value at the edge of non-integer values.
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float64RoundTrap2 = 4503599627370497.0
/// A round trap 64-bit float value at the edge of the exact integers
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float64RoundTrap3 = 9007199254740991.0;

/// Numerically smallest non-zero 32-bit float value.
const float32MinPositive = 1.401298464324817e-45;
/// Largest positive denormal 32-bit float value.
const float32MaxDenormal = 1.1754942106924411e-38;
/// Smallest positive normal (non-denormal) 32-bit float value.
const float32MinNormal = 1.1754943508222875e-38;
/// Largest 32-bit float value below 1.0.
const float32MaxFraction = 0.9999999403953552;
/// Largest non-integer 32-bit float value.
const float32MaxNonInteger = 8388607.5;
/// Largest exact 32-bit float integer value.
/// 
/// The largest integer 32-bit float where all smaller integers 
/// are also 32-bit floats.
/// All larger doubles represent ranges larger than one integer.
const float32MaxExactInteger = 16777216.0;
/// Largest finite 32-bit float value.
const float32MaxFinite = 3.4028234663852886e+38;
/// Positive infinity.
const float32Infinity = 1.0 / 0.0;
/// Not a number
const float32NaN = 0.0 / 0.0;
/// A round trap 32-bit float value at the edge of the precision of 1.0. 
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float32RoundTrap1 = 0.9999999403953552;
/// A round trap 32-bit float value at the edge of non-integer values.
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float32RoundTrap2 = 8388609.0
/// A round trap 32-bit float value at the edge of the exact integers
/// 
/// A "Round Trap" is a number where rounding to nearest by adding 0.5 and
/// rounding towards zero will not give the correct result.
const float32RoundTrap3 = 16777215.0;
```
